canonical/cos-lib
# added worker error on timeout if restart fails

Fixes #51 

## Issue
The worker attempts to start the services in the layer as soon as it has all the necessary pieces of config.
If that fails, for whatever reason, it currently reports 'active' and that's it. 
Any follow-up event, unless it signals config changes, will not even attempt to restart the services.

## Solution
The worker will retry for 15 minutes (default, configurable) to start the services and, if it fails, raise an exception and let the charm go into error state and write some logs that tell the user what's wrong and that it's probably an issue with some external services that the workload depends upon to start, such as s3.
The idea is that juju will retry the hook and eventually as those external services come up, the issue will resolve itself.

Rejected alternatives:
- attempting to 'gracefully' handle the situation by putting the charm in Blocked/Waiting are tricky because:
  - we become dependent on update-status or other events to wake us up so we can try again
  - we need to make the 'attempt to restart if not started' logic run unconditionally, which is an overcooked spaghetti code party
- using `resurrect` or something similar to wake yourself up with custom events periodically until the check finally passes
  - hard to predict how those will interact with regular juju events 

## Context
https://discourse.charmhub.io/t/its-probably-ok-for-a-unit-to-go-into-error-state/13022/30


## Testing Instructions
- Deploy a coordinator/worker charm built with this lib version
- Deploy an s3 facade so that the config will look good but s3 will in practice be broken
- Watch the charm cling to life for 15 minutes 
- Watch the charm go to error state
- Now replace the s3 facade for a real s3 and the charm should eventually go to active
